### ISSUE TITLE
leverage SIMD

### ISSUE THREAD

**Issue by LifeIsStrange** [plokhotnyuk/jsoniter-scala#650]
SIMD allow revolutionary intra-core parallelism. Actually the fastest Json library on earth is called simdjson for this precise reason.
openjdk 16 release next month and bring SIMD support to the JVM!
https://openjdk.java.net/jeps/338
You could hence use it that way (or through the intriguing https://github.com/beehive-lab/TornadoVM )

exciting isn't it ? :)
@plokhotnyuk

**Comment by plokhotnyuk** [comment]
I think that SIMD would work greatly only for payload with long strings.

A better option would be an adoption of SWAR techniques using 64-bit or 128-bit values, like [here](https://github.com/ibireme/yyjson).

**Comment by LifeIsStrange** [comment]
Wow interesting!!

**Comment by ScalaWilliam** [comment]
@plokhotnyuk that looks super super cool.

I am a little naive in C, still, what is the general approach the author is approaching, from a Scala perspective (if this is possible to explain at all)?

I did not know there was something that could be even faster than simdjson!

**Comment by plokhotnyuk** [comment]
@ScalaWilliam Before comparing the speed need to understand that the result of simdjson parsing is just an iterator over indexed JSON input, and not a data structure with arbitrary access to fields (values or references that accessible by offsets) as it usually happening in the Scala world. 

Sometime creation of the data structure in Scala takes more CPU cycles than parsing of included values from JSON. Most expensive are instances of immutable collections, boxed primitives and `Option[_]` wrappers.

**Comment by ScalaWilliam** [comment]
@plokhotnyuk thank you so much for your explanation - sorry I realised I did not say that intended to refer to `yyjson`. This library seems to extract everything out into immutable structures.

**Comment by plokhotnyuk** [comment]
@ScalaWilliam While `yyjson` allocates immutable nodes of the JSON object model the main CPU cycles will be spent during accessing to it when searching for tagged values that could be quite expensive for real world JSON messages.

Parsers from Scala world that provide object model for JSON usually use maps and vectors for JSON objects and JSON arrays accordingly that is much more expensive then parsing immediately to data structures and arrays.

**Comment by plokhotnyuk** [comment]
@LifeIsStrange @ScalaWilliam 

Latest versions of jsoniter-scala-core for JVM use SWAR techniques for parsing and serialization of ASCII strings, booleans, numbers, and `java.time._` values. 

It gives speed up to x2 for some cases.

Latest results are published here, as usually: https://plokhotnyuk.github.io/jsoniter-scala/

**Comment by plokhotnyuk** [comment]
@LifeIsStrange @ScalaWilliam Currently, jsoniter-scala seems to be quite competitive with simdjson-java when full validation against the JSON spec is not required for skipped keys and values: https://github.com/simdjson/simdjson-java/pull/2

**Comment by plokhotnyuk** [comment]
Here are results of benchmarks for jsoniter-scala Vs simdjson-java (parsing of different sizes of strings and arrays of booleans and numbers, tested on JDK-24).

TLDR: simdjson-java outperforms only in parsing of long ASCII strings: https://github.com/plokhotnyuk/jsoniter-scala/compare/master...simdjson#diff-f4916c36d9b86fe29ebd1ca34d19acb6b834dc1fecf5149c5cb319ee4bca2919R1

Also, simdjson-java is faster in skipping of JSON input that is not going to be decoded to some fields of resulting product-type instance: https://github.com/simdjson/simdjson-java?tab=readme-ov-file#512-bit-vectors

When using Oracle GraalVM JDK results for jsoniter-scala are better at ~10%, but simdjson-java slows down in 100x times or more because incubating JDK extension for vectors is not supported in GraalVM yet.
